test/FunctionalTests/Microsoft.OData.Core.Tests/IntegrationTests/Evaluation/AutoComputePayloadMetadataInJsonIntegrationTests.cs

```diff
@@ -1356,6 +1375,172 @@ private static IEdmModel GetModelWithAbstractType(out EdmEntitySet entitySet)
             return model;
         }
 
+        [Theory]
+        [InlineData("minimal")]
+        [InlineData("full")]
+        public void WritingContainedNavigationPropertyWithoutProvideKeyPropertyOnParentResourceWorks(string metadataLevel)
+        {
+            MemoryStream outputStream = new MemoryStream();
+            var container = ContainerBuilderHelper.BuildContainer(null);
+            IODataResponseMessage message = new InMemoryMessage() { Stream = outputStream, Container = container };
+
+            message.SetHeader("Content-Type", metadataLevel == "full" ? "application/json;odata.metadata=full" : "application/json");
+            ODataMessageWriterSettings settings = new ODataMessageWriterSettings();
+            var result = new ODataQueryOptionParser(Model, EntityType, EntitySet,
+                new Dictionary<string, string> { { "$select", "Name" }, { "$expand", "ContainedNavProp" } }).ParseSelectAndExpand();
+
+            ODataUri odataUri = new ODataUri()
+            {
+                ServiceRoot = new Uri("http://example.com"),
+                SelectAndExpand = result
+            };
+
+            Uri requestUri = new Uri("http://example.com/EntitySet(1)");
+            odataUri.RequestUri = requestUri;
+            odataUri.Path = new ODataUriParser(Model, new Uri("http://example.com"), requestUri).ParsePath();
+            settings.ODataUri = odataUri;
+
+            string output;
+            using (var messageWriter = new ODataMessageWriter(message, settings, Model))
+            {
+                ODataWriter writer = messageWriter.CreateODataResourceWriter(EntitySet, EntityType);
+
+                ODataResource resource = new ODataResource
+                {
+                    TypeName = "Namespace.EntityType",
+                    Properties = new[]
+                    {
+                        new ODataProperty { Name = "Name", Value = "SampleName" }
+                    }
+                };
+
+                if (metadataLevel == "full")
+                {
+                    resource.Id = new Uri("http://example.com/EntitySet(1)");

```

Comment:
what happens if metadata=full and we don't set this? Do we get the exception, or just omit? Is there any change to the old behavior for full metadata?

Reply:
It throws an exception. It's not for contained navigation property. If we only serialize a top level resource without "id" in full metadata, it also throws the same exception.